### PR TITLE
#529 allow for dynamic definition of columns

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
via reset property (@FabienLelaquais do you have/want a better name ?)

Example: switch column order
```
from taipy.gui import Gui

data = {"Col 1": [f"name {i}" for i in range(1000)], "Col 2": [True if i % 2 else False for i in range(1000)]}

columns="Col 1;Col 2"

def switch_cols(state):
    state.columns = "Col 2;Col 1" if state.columns == "Col 1;Col 2" else "Col 1;Col 2"

page = """
<|{data}|table|reset|columns={columns}|>
<|Switch Cols|button|on_action=switch_cols|>
"""

Gui(page=page).run()
```

### CODE REVIEW

File: src/taipy/gui/utils/_variable_directory.py [hunk 1]
```diff
@@ -103,7 +103,7 @@ def get_var(self, name: str, module: str) -> t.Optional[str]:
 
 
 _MODULE_NAME_MAP: t.List[str] = []
-_RE_TPMDL_DECODE = re.compile(r"(.*?)_TPMDL_(.*)")
+_RE_TPMDL_DECODE = re.compile(r"(.*?)_TPMDL_(\d+)$")
```
Comment (dinhlongviolin1) [inline src/taipy/gui/utils/_variable_directory.py]: Nice touch

File: src/taipy/gui/renderers/utils.py [hunk 2]
```diff
@@ -27,6 +27,10 @@ def _get_tuple_val(attr: tuple, index: int, default_val: t.Any) -> t.Any:
     return attr[index] if len(attr) > index else default_val
 
 
+def _get_column_desc(columns: t.Dict[str, t.Any], key: str) -> t.Dict[str, t.Any]:
```
Comment (dinhlongviolin1) [inline src/taipy/gui/renderers/utils.py]: Much needed refactor

**Review by dinhlongviolin1 — APPROVED** [review]
Nicely done

File: src/taipy/gui/utils/_variable_directory.py [hunk 3]
```diff
@@ -103,7 +103,8 @@ def get_var(self, name: str, module: str) -> t.Optional[str]:
 
 
 _MODULE_NAME_MAP: t.List[str] = []
-_RE_TPMDL_DECODE = re.compile(r"(.*?)_TPMDL_(.*)")
+_RE_HAS_TPMDL = re.compile(r".*_TPMDL_.*")
```
Comment (FabienLelaquais) [inline src/taipy/gui/utils/_variable_directory.py]: I don't see _RE_HAS_TPMDL anywhere... and the RE seems overkill.
If you need to find the string there are more effective ways

File: src/taipy/gui/renderers/builder.py [hunk 4]
```diff
@@ -464,44 +464,40 @@ def __update_col_desc_from_indexed(self, columns: t.Dict[str, t.Any], name: str)
...
-        col_types = self.__gui._accessors._get_col_types(data_hash, _TaipyData(data, data_hash))
-        columns = _get_columns_dict(
-            data,
-            _add_to_dict_and_get(self.__attributes, "columns", {}),
-            col_types,
-            _add_to_dict_and_get(self.__attributes, "date_format", date_format),
-            _add_to_dict_and_get(self.__attributes, "number_format", number_format),
-        )
+    def _get_dataframe_attributes(self, date_format="MM/dd/yyyy", number_format=None) -> '_Builder': # noqa: C901
+        columns, col_types, reset_fn_hash = self.__gui._calculate_table_columns(_is_boolean_true(t.cast(bool, self.__attributes.get("reset", False))), self.__hashes.get("reset"),
```
Comment (FabienLelaquais) [inline src/taipy/gui/renderers/builder.py]: Couldn't spot where _calculate_table_columns() is defined... strange

**Review by FabienLelaquais — APPROVED** [review]
I get the excellent idea.
Great job!

**Comment by FabienLelaquais** [comment]
> via reset property (@FabienLelaquais do you have/want a better name ?)
> 
> Example: switch column order
> 
> ```
> from taipy.gui import Gui
> 
> data = {"Col 1": [f"name {i}" for i in range(1000)], "Col 2": [True if i % 2 else False for i in range(1000)]}
> 
> columns="Col 1;Col 2"
> 
> def switch_cols(state):
>     state.columns = "Col 2;Col 1" if state.columns == "Col 1;Col 2" else "Col 1;Col 2"
> 
> page = """
> <|{data}|table|reset|columns={columns}|>
> <|Switch Cols|button|on_action=switch_cols|>
> """
> 
> Gui(page=page).run()
> ```

How about "reconstruct"?

Please create another issue for me to create a full example of this, and update the ReleaseNotes

File: src/taipy/gui/utils/_variable_directory.py (hunk 3 above)
Comment (FredLL-Avaiga) [reply src/taipy/gui/utils/_variable_directory.py]: it is used in Gui but you're right it is an overkill

File: src/taipy/gui/renderers/builder.py (hunk 4 above)
Comment (FredLL-Avaiga) [reply src/taipy/gui/renderers/builder.py]: in Gui

**Comment by FredLL-Avaiga** [comment]
> How about "reconstruct"?
> 
Don't like it

**Pull request merged**